File: counting/Counting.java
import java.util.Arrays;
import java.util.Random;

public class Counting {
    public static void main(String[] args) {
        Random random = new Random();
        int[] lista1 = random.ints(10201, -100, 100000).toArray();

        int[] lista2 = Arrays.copyOf(lista1, lista1.length);
        int[] ordenado = outPlace(lista1);
        inPlace(lista2);

        assert Arrays.equals(ordenado, lista2);
        assert estaOrdenado(lista2);
    }


    public static void inPlace(int[] a) {
        int[] mm = maiorMenor(a);
        int maior = mm[0];
        int menor = mm[1];

        int[] c = new int[maior - menor + 1];
        for (int i = 0; i < a.length; i++) {
            c[a[i] - menor]++;
        }

        for (int i = 1; i < c.length; i++) {
            c[i] += c[i - 1];
        }

        int[] b = Arrays.copyOf(a, a.length);
        for (int i = b.length - 1; i >= 0; i--) {
            a[c[b[i] - menor] - 1] = b[i];
            c[b[i] - menor]--;
        }
    }


    public static int[] outPlace(int[] a) {
        int[] mm = maiorMenor(a);
        int maior = mm[0];
        int menor = mm[1];

        int[] c = new int[maior - menor + 1];
        for (int i = 0; i < a.length; i++) {
            c[a[i] - menor]++;
        }

        for (int i = 1; i < c.length; i++) {
            c[i] += c[i - 1];
        }

        int[] b = new int[a.length];
        for (int i = a.length - 1; i >= 0; i--) {
            b[c[a[i] - menor] - 1] = a[i];
            c[a[i] - menor]--;
        }

        return b;
    }


    private static int[] maiorMenor(int[] a) {
        if (a.length == 0) return new int[] {0, 0};

        int maior = a[0];
        int menor = a[0];

        for (int i = 0; i < a.length; i++) {
            if (a[i] > maior) maior = a[i];
            if (a[i] < menor) menor = a[i];
        }
        return new int[] {maior, menor};
    }


    private static boolean estaOrdenado(int[] v) {
        for (int i = 0; i < v.length - 1; i++) {
            if (v[i] > v[i + 1]) return false;
        }
        return true;
    }
}

File: counting/counting.py
from random import randrange
from itertools import repeat

def maior_menor(v):
    if len(v) == 0:
        return 0, 0

    maior = v[0]
    menor = v[0]

    for e in v:
        if e > maior:
            maior = e
        if e < menor:
            menor = e

    return maior, menor


def out_place(a):
    maior, menor = maior_menor(a)
    c = [0] * (maior - menor + 1)

    for i in range(len(a)):
        c[a[i] - menor] += 1

    for i in range(1, len(c)):
        c[i] += c[i - 1]

    b = [0] * len(a)
    for i in range(len(a) - 1, -1, -1):
        b[c[a[i] - menor] - 1] = a[i]
        c[a[i] - menor] -= 1

    return b


def in_place(a): # in-place no sentido que altera os elementos no próprio array, é impossível fazer um realmente in-place, possivelmente daria pra não usar um array auxiliar sem ser o de contagem, enfim
    maior, menor = maior_menor(a)
    c = [0] * (maior - menor + 1)

    for i in range(len(a)):
        c[a[i] - menor] += 1

    for i in range(1, len(c)):
        c[i] += c[i - 1]

    b = a.copy()

    for i in range(len(b) - 1, -1, -1):
        a[c[b[i] - menor] - 1] = b[i]
        c[b[i] - menor] -= 1

if __name__ == '__main__':
    lista1 = [randrange(100000) for _ in repeat(None, 100)] + [-1000]

    lista2 = out_place(lista1)
    print(lista2)

    in_place(lista1)
    print(lista1)

    assert lista1 == list(sorted(lista1)) == lista2